### PR TITLE
[GPU] Slice operation

### PR DESCRIPTION
**Pull request by ytorzuk-altran**
### Details:
 - Slice operation in cldnn_engine was implemented according to https://github.com/openvinotoolkit/openvino/blob/master/docs/ops/movement/Slice_8.md
 - Slice functionality is covered by single layer tests

Ticket: *54648*

Due to many changes in GPU code structure this PR was rebased and reopened here: https://github.com/openvinotoolkit/openvino/pull/9142

### CODE REVIEW

**Review by vladimir-paramuzov — APPROVED** [review]

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/slice/slice_kernel_ref.cpp [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+JitConstants SliceKernelRef::GetJitConstants(const slice_params &params) const {
+    JitConstants jit = MakeBaseParamsJitConstants(params);
+    addJitConstantsForAttribute(jit, "SLICE_BEGIN", params.start);
+    addJitConstantsForAttribute(jit, "SLICE_END", params.end);
+    addJitConstantsForAttribute(jit, "SLICE_STEP", params.step);
+    return jit;
+}
+
+CommonDispatchData SliceKernelRef::SetDefault(const slice_params &params,
```
Comment (tgubanova-lohika) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/slice/slice_kernel_ref.cpp:95]: There is no need to be a member of class (at least it make sense to make the method static).
Also "SetDefault" is a little bit unclear name, maybe better to rename to `createCommonDispatchDate` or sth like this.

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/slice/slice_kernel_ref.cpp [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    if (params.inputs.empty())
+        return false;
+
+    if (params.output.Dimentions() > 5 || params.inputs[0].Dimentions() > 5)
+        return false;
+
+    return true;
+}
+
+JitConstants SliceKernelRef::GetJitConstants(const slice_params &params) const {
```
Comment (tgubanova-lohika) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/slice/slice_kernel_ref.cpp:87]: this method also can be static, isn't it?

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/slice/slice_kernel_ref.cpp (hunk 1 above)
Comment (ytorzuk-altran) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/slice/slice_kernel_ref.cpp:95]: There is a [request](https://github.com/openvinotoolkit/openvino/pull/8267#discussion_r738241506) from Vladimir to align names with other kernels

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/slice/slice_kernel_ref.cpp (hunk 2 above)
Comment (ytorzuk-altran) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/slice/slice_kernel_ref.cpp:87]: There is a [request](https://github.com/openvinotoolkit/openvino/pull/8267#discussion_r738241506) from Vladimir to align names with other kernels

Comment (tgubanova-lohika) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/slice/slice_kernel_ref.cpp:87]: but they can be static, right?

**Comment by vladimir-paramuzov** [comment]
@ytorzuk-altran Could you rebase on top of latest master and resolve conflicts? I will merge PR once it has green CI

**Comment by ytorzuk-altran** [comment]
> @ytorzuk-altran Could you rebase on top of latest master and resolve conflicts? I will merge PR once it has green CI

@vladimir-paramuzov It turns out that Slice-8 op is already enabled by c171be238fa920bfcf86e3d46399d05e962ac32a in the current master by replacing Slice-8 with StridedSlice in transformations. Though, there are a few SL tests from this PR which fail when running them on the Slice-8 version from master.
Considering the latest changes in master, do you think it makes sense to have a dedicated Slice-8 op instead of implementation based on StridedSlice?

**Comment by vladimir-paramuzov** [comment]
@mitruska Does Slice -> StridedSlice conversion in ngraph cover ALL possible parameters/shapes of Slice operation or some limitation exist?

**Comment by mitruska** [comment]
> @mitruska Does Slice -> StridedSlice conversion in ngraph cover ALL possible parameters/shapes of Slice operation or some limitation exist?

The main limitation for the Slice -> StridedSlice transformation is that `axes` input must be Constant 
@ytorzuk-altran Could you precise which cases are failing with master version?

**Comment by ytorzuk-altran** [comment]
> > @mitruska Does Slice -> StridedSlice conversion in ngraph cover ALL possible parameters/shapes of Slice operation or some limitation exist?
> 
> The main limitation for the Slice -> StridedSlice transformation is that `axes` input must be Constant @ytorzuk-altran Could you precise which cases are failing with master version?

@mitruska Here you are.
inference-engine/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/slice.cpp:17
inference-engine/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/slice.cpp:28

The root cause of both failures is an issue in `start` parameter normalization in case passed `start` parameter is out of the input range.

**Comment by ytorzuk-altran** [comment]
@mitruska Luckly, SLTs from this PR are fully compatible with the `SliceLayerTest` you've implemented.

**Comment by mitruska** [comment]
> @mitruska Luckly, SLTs from this PR are fully compatible with the `SliceLayerTest` you've implemented.

Yes, I have included your test cases in the CPU SLT to make sure there are compatible.

So the cases that you've mentioned are passing on current master with CPU, 
seems it can be a problem with GPU StridedSlice.

GPU (Current PR):
https://github.com/openvinotoolkit/openvino/blob/73547acb01c622a3c261b5705ad3c579a331a8cf/inference-engine/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/slice.cpp#L17
https://github.com/openvinotoolkit/openvino/blob/73547acb01c622a3c261b5705ad3c579a331a8cf/inference-engine/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/slice.cpp#L28

Corresponding CPU (master):
https://github.com/openvinotoolkit/openvino/blob/9972410aa39977c0f7ab6edfc7da1863532a9e73/inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/slice.cpp#L29
https://github.com/openvinotoolkit/openvino/blob/9972410aa39977c0f7ab6edfc7da1863532a9e73/inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/slice.cpp#L40

**Comment by ytorzuk-altran** [comment]
@vladimir-paramuzov This PR was rebased onto the latest master. ConvertSliceToStridedSlice transformation was disabled in GPU plugin in favor of GPU Slice-8 implementation.

File: inference-engine/src/cldnn_engine/cldnn_transformations_pipeline.cpp [hunk 3]
```diff
@@ -308,6 +309,7 @@ void TransformationsPipeline::apply(std::shared_ptr<ov::Function> func) {
         pass_config->disable<ngraph::pass::WeightsDequantizeToFakeQuantize>();
         pass_config->disable<ngraph::pass::SimplifyCTCGreedyDecoderSeqLen>();
         pass_config->enable<ngraph::pass::ConvertGather8ToGather7>();
+        pass_config->disable<ngraph::pass::SliceToStridedSlice>();
```
Comment (GlebKazantaev) [inline inference-engine/src/cldnn_engine/cldnn_transformations_pipeline.cpp:312]: This is not going to work as this transformation is located inside MOC so we execute it in offline and in runtime.
Also the main reason why it is located inside MOC is that Slice is participating in common optimizations which works with more generic way of slice which is StridedSlice operation, so we have to align all slice-type operations to the same type (SS).
So if you have more efficient implementation for Slice you have to write StridedSliceToSlice conversion and enable it inside GPU plugin. If more than one plugin have efficient implementation for Slice we can enable StridedSliceToSlice conversion inside CommonOptimizations pipeline.

**Review by GlebKazantaev — CHANGES_REQUESTED** [review]

**Comment by ytorzuk-altran** [comment]
* The branch was rebased onto the latest master
* ConvertSliceToStridedSlice transformation disabling was dropped
* SLTs were dropped because StridedSlice-based implementation is preferable for now
* Added cldnn unit  tests which cover Slice functionality

File: inference-engine/src/cldnn_engine/cldnn_transformations_pipeline.cpp (hunk 3 above)
Comment (ytorzuk-altran) [reply inference-engine/src/cldnn_engine/cldnn_transformations_pipeline.cpp]: ConvertSliceToStridedSlice transformation disabling was dropped

**Comment by lznamens** [comment]
@ytorzuk-altran could you please fix segfault as @sshlyapn describes in this PR: https://github.com/openvinotoolkit/openvino/pull/9064

@yeonbok Please duplicate your comments from https://github.com/openvinotoolkit/openvino/pull/9142 into this PR. And I'll close 9142 as unneeded.

**Comment by tgubanova-lohika** [comment]
> @ytorzuk-altran could you please fix segfault as @sshlyapn describes in this PR: #9064
> 
> @yeonbok Please duplicate your comments from #9142 into this PR. And I'll close 9142 as unneeded.

@lznamens We agreed with Pavel Durandin to continue contributing to #9142 and close this one on Tuesday

**Comment by lznamens** [comment]
> > @ytorzuk-altran could you please fix segfault as @sshlyapn describes in this PR: #9064
> > @yeonbok Please duplicate your comments from #9142 into this PR. And I'll close 9142 as unneeded.
> 
> @lznamens We agreed with Pavel Durandin to continue contributing to #9142 and close this one on Tuesday

I got it, thanks.

**Comment by ytorzuk-altran** [comment]
@lznamens @yeonbok @tgubanova-lohika I'm unable to modify branch @lznamens has created, hence I've force-pushed rebased changes to the branch associated with this PR. These changes include modifications according to @yeonbok comments under #9142.

**Review by ilyachur — APPROVED** [review]
nGraph part LGTM

**Pull request merged**